ambry-commons/src/main/java/com/github/ambry/commons/HostLevelThrottler.java

```diff
@@ -0,0 +1,101 @@
+import java.util.HashMap;
+import java.util.Map;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A class to decide if a request should be throttled.
+ */
+public class HostLevelThrottler {
+  private static final Logger logger = LoggerFactory.getLogger(HostLevelThrottler.class);
+  private final Map<RestMethod, RejectThrottler> quotaMap;
+  private final Map<HardwareResource, Criteria> hardwareThresholdMap;
+  private final HardwareUsageMeter hardwareUsageMeter;
+
+  public HostLevelThrottler(HostThrottleConfig hostThrottleConfig) {
+    this.hardwareUsageMeter =
+        new HardwareUsageMeter(hostThrottleConfig.cpuMaxAllowedDelayInMs, hostThrottleConfig.memoryMaxAllowedDelayInMs);
+    JSONObject quota = new JSONObject(hostThrottleConfig.restRequestQuota);
+    JSONObject thresholds = new JSONObject(hostThrottleConfig.hardwareThresholds);
+    quotaMap = new HashMap<>();
+    hardwareThresholdMap = new HashMap<>();
+    for (RestMethod restMethod : RestMethod.values()) {
+      quotaMap.put(restMethod, new RejectThrottler(quota.optInt(restMethod.name(), -1)));
+    }
+    // All criteria should be satisfied.
+    // Future work: if multiple throttling rules under "OR" condition needed, a list of thresholds/hardwareThresholdMap can be used.
+    for (HardwareResource hardwareResource : HardwareResource.values()) {
+      if (thresholds.has(hardwareResource.name())) {
+        JSONObject jsonObject = thresholds.getJSONObject(hardwareResource.name());
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+          Criteria criteria = objectMapper.readValue(jsonObject.toString(), Criteria.class);
+          hardwareThresholdMap.put(hardwareResource, criteria);
+        } catch (IOException e) {
+          logger.error("Error while parsing {} threshold ", hardwareResource.name(), e);
+          continue;
+        }
+      }
+    }
+    logger.info("Host throttling config: {} {}", quotaMap, hardwareThresholdMap);
+  }
+
+  public HostLevelThrottler(Map<RestMethod, RejectThrottler> quotaMap,
+      Map<HardwareResource, Criteria> hardwareResourcesCriteriaMap, HardwareUsageMeter hardwareUsageMeter) {
+    this.quotaMap = quotaMap;
+    this.hardwareThresholdMap = new HashMap<>();
+    this.hardwareUsageMeter = hardwareUsageMeter;
+  }
+
+  /**
+   * Return {@code true} if throttling is required. Return {@code false} if no throttler for this rest method or quota
+   * is not reached.
+   * @param restRequest provides the information.
+   */
+  public boolean shouldThrottle(RestRequest restRequest) {
+    boolean shouldThrottle;
+    // Check quotaMap
+    RejectThrottler throttler = quotaMap.get(restRequest.getRestMethod());
+    shouldThrottle = (throttler != null && throttler.shouldThrottle(1));
+    // Check hardwareThresholdMap
+    if (!shouldThrottle && hardwareThresholdMap.containsKey(HardwareResource.MEMORY)) {
```

Comment:
A cleaner way to implement both the blocks could be iterating for all `HardwareResource`s. This will also help to add new hardware resources without code change here.

for(HardwareResource hardwareResource : HardwareResource.values()) {
     if (!shouldThrottle && hardwareThresholdMap.containsKey(hardwareResource)) {
        shouldThrottle = !hardwareThresholdMap.get(HardwareResource.MEMORY)
          .meetRequirement((long) hardwareUsageMeter.getMemoryPercentage());  // TODO (NOTE: get the specific resource requirement from a map)
       if (shouldThrottle) {
        logger.warn("Throttle because of Memory: {}%", hardwareUsageMeter.getMemoryPercentage());
      }
}

Reply:
+1

Reply:
fixed.